Sources/NIOSSL/SSLCallbacks.swift

```diff
@@ -173,6 +173,108 @@ public struct PSKClientIdentityResponse: Sendable {
     }
 }
 
+/// A structure representing values from client extensions in the SSL/TLS handshake.
+///
+/// This struct contains values obtained from the client hello message extensions during the TLS handshake process and
+/// can be manipulated or introspected by the `NIOSSLContextCallback` to alter the TLS handshake behaviour dynamically
+/// based on these values.
+public struct NIOSSLClientExtensionValues: Hashable {
+
+    /// The hostname value from the Server Name Indication (SNI) extension.
+    ///
+    /// This value, if available, indicates the requested server hostname by the client.
+    /// In a context where a service is handling multiple hostnames (virtual hosts, for example),
+    /// this value could be used to decide which SSLContext to use for the handshake.
+    public var serverHostname: String?
+
+    /// Initializes a new `NIOSSLClientExtensionValues` struct.
+    ///
+    /// - parameter serverHostname: The hostname value from the SNI extension.
+    public init(serverHostname: String?) {
+        self.serverHostname = serverHostname
+    }
+}
+
+/// A callback that can used to support multiple or dynamic TLS hosts.
+///
+/// When set, this callback will be invoked once per TLS hello. The provided `NIOSSLClientExtensionValues` will contain the
+/// host name indicated in the TLS client hello, while the provided `NIOSSLContext` will be the current
+/// server context being used for the handshake.
+///
+/// Within this callback, the user can create and return a new server `NIOSSLContext` for the given host,
+/// return the same context provided if it is sufficient to complete the handshake, or return `throw` in the event
+/// of an exception.
+///
+public typealias NIOSSLContextCallback = @Sendable (NIOSSLClientExtensionValues, EventLoopPromise<NIOSSLContext>) -> Void
```

Comment:
I increasingly agree with you @AndrewBarba that this is maybe a bit too big. You proposed returning a new TLSConfiguration object instead and applying the delta. That's a nice idea, but I wonder if we should instead define a third object (maybe a "post extension config change" object) with a trimmed-down set of fields. That would let us extend it in the future, and avoids some config that would be very hard for us to change. We could start out only with the cert chain.

What do you think?

Reply:
@Lukasa Big +1 from me on this idea. I really like a dedicated struct with just the deltas we want to support. Going to give this a shot later this week

Reply:
Fantastic, please let me know when that's done and I'll do another review pass. I'm trying to minimise the number of incremental reviews I do this week, because it's WWDC week and I'm slammed.